jsonrpc/server.go

```diff
@@ -195,10 +195,21 @@ func (s *Server) Handle(ctx context.Context, data []byte) ([]byte, error) {
 	return s.HandleReader(ctx, bytes.NewReader(data))
 }
 
+type connKey struct{}
+
+func ConnFromContext(ctx context.Context) (io.ReadWriter, bool) {
+	// We don't need to worry about conn being nil since we
+	// store the connection in the context on every request.
+	conn, ok := ctx.Value(connKey{}).(io.ReadWriter)
+	return conn, ok
+}
+
 // HandleReader processes a request to the server
 // It returns the response in a byte array, only returns an
 // error if it can not create the response byte array
 func (s *Server) HandleReader(ctx context.Context, reader io.Reader) ([]byte, error) {
+	connCtx := context.WithValue(ctx, connKey{}, reader)
```

Comment:
I think it makes sense to move this to transport side (ie http, ws) That way we can fix #1271 pretty easily.